optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/common/accessor/gizmo/GizmoMemberDescriptor.java

```diff
@@ -0,0 +1,237 @@
+     * and the first letter lowercase; otherwise, the method name.
+     */
+    String name;
+
+    /**
+     * If the member is a field, the FieldDescriptor of the member
+     * If the member is a method, the MethodDescriptor of the member
+     */
+    Object memberDescriptor;
+
+    /**
+     * If the member is a normal member, the class that declared it
+     * If the member is from Jandex, the Jandex ClassInfo of the class that declared it
+     */
+    Class<?> declaringClass;
+
+    /**
+     * The member as an AnnotatedElement
+     */
+    AnnotatedElement annotatedElement;
+
+    /**
+     * The MethodDescriptor of the corresponding setter. Is empty if not present.
+     */
+    Optional<MethodDescriptor> setter;
+
+    /**
+     * The generic type of the member
+     */
+    Type type;
+
+    public GizmoMemberDescriptor(Member member) {
+        declaringClass = member.getDeclaringClass();
+        if (!Modifier.isPublic(member.getModifiers())) {
+            throw new IllegalStateException("Member (" + member.getName() + ") of class (" +
+                    member.getDeclaringClass().getName() + ") is not public and domainAccessType is GIZMO.\n" +
+                    ((member instanceof Field) ? "Maybe put the annotations onto the public getter of the field.\n" : "") +
+                    "Maybe use domainAccessType REFLECTION instead of GIZMO.");
+        }
+        if (member instanceof Field) {
+            memberDescriptor = FieldDescriptor.of((Field) member);
+            name = member.getName();
+            annotatedElement = (Field) member;
+            type = ((Field) member).getGenericType();
+            setter = lookupSetter();
+        } else if (member instanceof Method) {
+            memberDescriptor = MethodDescriptor.ofMethod((Method) member);
+            annotatedElement = (Method) member;
+            if (ReflectionHelper.isGetterMethod((Method) member)) {
+                name = ReflectionHelper.getGetterPropertyName(member);
+            } else {
+                name = member.getName();
+            }
+            type = ((Method) member).getGenericReturnType();
+            setter = lookupSetter();
+        } else {
+            throw new IllegalArgumentException(member + " is not a Method or a Field.");
+        }
+    }
+
+    // For Quarkus
+    // (Cannot move to Quarkus module; get runtime
+    //  exception since objects created here use classes
+    //  from another ClassLoader).
+    public GizmoMemberDescriptor(String name, Object memberDescriptor, Class<?> declaringClass,
+            AnnotatedElement annotatedElement, Type type) {
+        this(name, memberDescriptor, declaringClass, annotatedElement, type, null);
+    }
+
+    public GizmoMemberDescriptor(String name, Object memberDescriptor, Class<?> declaringClass,
+            AnnotatedElement annotatedElement, Type type,
+            Optional<MethodDescriptor> setterDescriptor) {
```

Comment:
Sonar doesn't like this, and I agree. `Optional` method argument is a good sign that you need two methods instead. One without the argument, which could delegate to the one with the argument.

Reply:
Sonar doesn't like this because I use null, empty, and present (for the sake of having one constructor delegate to another; null means lookup, otherwise use directly the Optional). The Optional is the method descriptor of the setter; the one without the Optional is already taken (it for the lookup case). I could use Optional.ofNullable, but that involves unwrapping an already Optional value.